### PR TITLE
Implement source/line number support for Exception.StackTrace.

### PR DESCRIPTION
**Pull request by mikem8361**
Uses a new helper class (StackTraceSymbols) in System.Diagnostics.StackTrace assembly
to get source/line info for portable PDBs. On Windows both types of pdbs are support;
on xplat only portable PDBs.

The first part of the work is in the GetStackFramesInternal pinvoke to return the neccessary
info needed by the portable pdb reader: MethodToken, assembly path, loaded PE image address
and if in-memory symbol file, the address.

The second part is in the StackFrameHelper in mscorlib. It attempts to load the System.Diagnostics.StackTrace,
gets the new type and creates a delegate to the helper function which is called for each stack
frame.

### CODE REVIEW

**Comment by mikem8361** [comment]
Depends on PR https://github.com/dotnet/corefx/pull/8670 in corefx.

**Comment by mikem8361** [comment]
@ellismg, @stephentoub @jkotas @noahfalk I addressed all the code review feedback.

1) removed all the catch-all try/catches
2) added some recursion protection
3) updated to the proper coding convention
4) moved the code in the new assembly to System.Diagnostics.StackTrace
5) added caching of metadata readers using the loaded PE image address
6) misc cleanup.

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 1]
```diff
@@ -74,10 +98,81 @@ public StackFrameHelper(bool fNeedFileLineColInfo, Thread target)
...
+            // Check if this function is being reentered because of an exception in the code below
+            if (t_reentrancy > 0)
+                return;
+
+            t_reentrancy++;
+            try
+            {
+                if (s_symbolsMethodInfo == null)
+                {
+                    Assembly metadataAssembly = Assembly.Load("System.Diagnostics.StackTrace, Version=1.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a");
```
Comment (jkotas) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: `Assembly.Load` method never returns null - it always throws.

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 2]
```diff
@@ -74,10 +98,81 @@ public StackFrameHelper(bool fNeedFileLineColInfo, Thread target)
...
+                    if (metadataAssembly == null)
+                        return;
+
+                    s_symbolsType = metadataAssembly.GetType("System.Diagnostics.StackTraceSymbols");
```
Comment (jkotas) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: Same here.

You may want to combine the both calls into:

`Type.GetType("System.Diagnostics.StackTraceSymbols, System.Diagnostics.StackTrace...", throwOnError: false)`

Comment (jkotas) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: `throwOnError: false` will catch the "right" errors for you

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 3]
```diff
@@ -74,10 +98,81 @@ public StackFrameHelper(bool fNeedFileLineColInfo, Thread target)
...
+
+                    s_symbolsMethodInfo = s_symbolsType.GetMethod("GetSourceLineInfo");
+                    if (s_symbolsMethodInfo == null)
+                        return;
+                }
+
+                if (getSourceLineInfo == null)
+                {
+                    // Create an instance of System.Diagnostics.Stacktrace.Symbols
+                    object target = s_symbolsType.GetConstructor(new Type[] { }).Invoke(null);
```
Comment (jkotas) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: `Activator.CreateInstance(s_symbolsType)` is more common way to create instance of the object with default constructor.

**Comment by jkotas** [comment]
LGTM module comments.

**Pull request merged**